### DISCUSSION TITLE
Support for EQ3 BLE thermostat

### DISCUSSION THREAD

**Discussion by Schwenkner** [arendst/Tasmota#10919] [Q&A] [answered]
I'm currently trying to implement support for the EQ3 thermostats from EQ-3.
There are some of them having BLE. It could be done for tasmota32.

I have currently a kind of draft code within tasmota, which is able to set the demanded temperature. This is only for the understanding, how it could be integrated into tasmota. The  remaining functions can then be implemented analog. Since I'm very beginning to understand how tasmota is working, I have a lot of questions.

Is there someone from the "Core-Team" who is willing to answer some questions I have and help me to finalize the draft.
Some basics I have figured out, but I assume it cone be done much better.

So, if there is someone interested in, pls let a message here.

**Comment by Jason2866** [comment, marked as answer]
Have you tried this? https://github.com/btsimonh/Tasmota/tree/EQ3_2020-02-04

**Reply by Schwenkner** [reply]
Not yet, thx, I'll give it a try. Have had a look only into master branch and didn't find it there.
With a first  fast look I have seen, that there is no scan for EQ3 devices or did I miss it.

**Comment by Schwenkner** [comment]
So, I've had some time to play with this branch. Unfortunately I coudln't get it working. This branch is much more sphisticated then my draft and it integrates in the common BLE implementation which mine currently doesn't.

First it didn't start scanning at all. I found out, that
BLEMasterEnable = Settings.flag5.mi32_enable;
was "false". Since I dit not know how to set "Settings.flag5.mi32_enable" this to true, I've done this directly.
Then I've got scan results. But the scans for EQ3 have been empty.

To analize this, I've added this into "BLEscanEndedCB" of "xdrv_79_esp32_ble.ino" which seemed to me as the place I have all scan and then forwarded to the scansubscribers

    for (uint32_t i = 0; i < results.getCount(); i++)
      {
        NimBLEAdvertisedDevice advertisedDevice = results.getDevice(i);
        AddLog(LOG_LEVEL_INFO,PSTR("BLE: scan compelete %s"), advertisedDevice.toString().c_str());
      }

Here I get only two scan resulst;
scan compelete Name: , Address: 7a:fb:af:e7:87:82, serviceUUID: 0xfeaa
scan compelete Name: , Address: 30:f7:72:57:c8:4a, manufacturer data: .....

Doing the same in my implementation, I get a lot more advertisements, about 18. This first address is not amoung them, the second was
Why do I get only two advertisment here? Do I missing again something like the Settings.flag5?

Then I have some traces in "TaskEQ3advertismentCallback". I have seen, it is checked for a special device name. I'm searching the EQ3 devices not by name, but by serviceUUID. With the traces I would check, if eventually the name doesn't match. But what I see here is only the advertisement with manufacturer data, not the one with the serviceUUID. If the advertisement with the serviceUUID is nto forwarded to EQ3, then how these device can be sanned ar all.

Does the implementation of this branch somehow stop to scan? Can I prevent this?

BTW, now I've got 4 advertisments two with manufaturer data, two with a serviceUUID. But none of them match with the EQ3 devive and none of them match with any addresses my application shows. I'me totally confused.

**Comment by Schwenkner** [comment]
Some time later I figured out, why tasmota stops to scan.

Settings.flag5.mi32_enable is the reason. If this is not propperly set, then it stops after one second.
How can I set this to "true"?

And now I also get my EQ3 (and my be the one from my neighbour) in the devlist, "Hurra".
Let's see, what he will do, if I change his settings.

**Reply by barbudor** [reply]
There is a lot of on-going activity on ESP32 BLE.
You should reach the Tasmota Discord server on the #ble-esp32 channel as the people involve not necesseraly watch Github discussions.

**Reply by Schwenkner** [reply]
Thx. I'm new to discord. Do I need an invitation for that channel? How do I get it? Cannot figure out how to participate without invitation.

**Comment by Jason2866** [comment]
Click [DISCORD Chat](https://discord.gg/Ks2Kzd4)
